modules/fedimint-ln-client/src/lib.rs

```diff
@@ -1357,6 +1358,79 @@ impl LightningClientModule {
 
+    /// Scan unspent incoming contracts for a payment hash that matches a
+    /// tweaked keys in the `indices` vector
+    pub async fn scan_receive_for_user_tweaked(
+        &self,
+        key_pair: KeyPair,
+        indices: Vec<u64>,
+    ) -> anyhow::Result<Amount> {
+        for i in indices {
+            let key_pair_tweaked = tweak_user_secret_key(&self.secp, key_pair, i);
+            if let Ok(amount) = self.scan_receive_for_user(key_pair_tweaked).await {
+                return Ok(amount);
+            }
+        }
+
+        Err(anyhow::anyhow!("No contract found for tweaked keys"))
+    }
+
+    /// Scan unspent incoming contracts for a payment hash that matches a public
+    /// key and claim the incoming contract
+    pub async fn scan_receive_for_user(&self, key_pair: KeyPair) -> anyhow::Result<Amount> {
+        let preimage_key: [u8; 33] = key_pair.public_key().serialize();
+        let preimage = sha256::Hash::hash(&preimage_key);
+        let contract_id = ContractId::from_hash(sha256::Hash::hash(&preimage));
+        self.claim_funded_incoming_contract(
+            key_pair,
+            contract_id,
+            "claiming unspent funded incoming contract",
+        )
+        .await
+    }
+
+    /// Claim the funded, unspent incoming contract by submitting a transaction
+    /// to the federation and awaiting the primary module's outputs
+    async fn claim_funded_incoming_contract<M: Serialize + Send + Sync>(
+        &self,
+        key_pair: KeyPair,
+        contract_id: ContractId,
+        extra_meta: M,
+    ) -> anyhow::Result<Amount> {
+        let incoming_contract_account = get_incoming_contract(self.module_api.clone(), contract_id)
+            .await?
+            .ok_or(anyhow::anyhow!("No contract account found"))
+            .with_context(|| format!("No contract found for {contract_id:?}"))?;
+
+        let input = incoming_contract_account.claim();
+        let client_input = ClientInput::<LightningInput, LightningClientStateMachines> {
+            input,
+            keys: vec![key_pair],
+            state_machines: Arc::new(|_, _| vec![]),
+        };
+
+        let tx =
+            TransactionBuilder::new().with_input(self.client_ctx.make_client_input(client_input));
+        let extra_meta = serde_json::to_value(extra_meta).expect("extra_meta is serializable");
+        let operation_meta_gen = |_, _| LightningOperationMeta {
+            variant: LightningOperationMetaVariant::Claim,
+            extra_meta: extra_meta.clone(),
+        };
+        let operation_id = OperationId::new_random();
+        let (_txid, out_points) = self
+            .client_ctx
+            .finalize_and_submit_transaction(
+                operation_id,
+                LightningCommonInit::KIND.as_str(),
+                operation_meta_gen,
+                tx,
+            )
+            .await?;
+        self.client_ctx
+            .await_primary_module_outputs(operation_id, out_points)
+            .await
```

Comment:
Waiting here makes the function blocking, why do that? This fn should return an operation id and there should be a `subscribe_*_updates` fn instead that can be used to follow the progress of the claim.

I see the update stream was changed to accomodate such claims, so should already work.

Reply:
I figured this function would already be called in the background, since the wallet will need to listen (or poll) the LNURL server when it receives. Returning the `OperationId` works for me too, I will change it to that

Reply:
Changed to non-blocking